src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonPatternDetector.cs

```diff
@@ -0,0 +1,307 @@
+                            strict = true;
+                        }
+                        else
+                        {
+                            break;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static bool IsMethodArgument(SyntaxToken token, ISyntaxFactsService syntaxFacts)
+            => syntaxFacts.IsLiteralExpression(token.Parent) &&
+               syntaxFacts.IsArgument(token.Parent.Parent) &&
+               syntaxFacts.IsInvocationExpression(token.Parent.Parent.Parent.Parent);
+
+        public bool IsDefinitelyJson(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (IsDefinitelyNotJson(token, _syntaxFacts))
+            {
+                return false;
+            }
+
+            if (HasJsonLanguageComment(token, _syntaxFacts, out _))
+            {
+                return true;
+            }
+
+            if (!IsMethodArgument(token, _syntaxFacts))
+            {
+                return false;
+            }
+
+            var stringLiteral = token;
+            var literalNode = stringLiteral.Parent;
+            var argumentNode = literalNode.Parent;
+            Debug.Assert(_syntaxFacts.IsArgument(argumentNode));
+
+            var argumentList = argumentNode.Parent;
+            var invocationOrCreation = argumentList.Parent;
+            if (_syntaxFacts.IsInvocationExpression(invocationOrCreation))
+            {
+                var invokedExpression = _syntaxFacts.GetExpressionOfInvocationExpression(invocationOrCreation);
+                var name = GetNameOfInvokedExpression(invokedExpression);
+                if (_syntaxFacts.StringComparer.Equals(name, _methodNameOfInterest))
+                {
+                    // Is a string argument to a method that looks like it could be a Regex method.  
+                    // Need to do deeper analysis
+                    var method = _semanticModel.GetSymbolInfo(invocationOrCreation, cancellationToken).GetAnySymbol();
+                    if (method != null &&
+                        method.DeclaredAccessibility == Accessibility.Public &&
+                        method.IsStatic &&
+                        _typesOfInterest.Contains(method.ContainingType))
+                    {
+                        return AnalyzeStringLiteral(
+                            stringLiteral, argumentNode, cancellationToken);
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public JsonTree TryParseJson(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (IsDefinitelyNotJson(token, _syntaxFacts))
+            {
+                return null;
+            }
+
+            HasJsonLanguageComment(token, _syntaxFacts, out var strict);
+
+            var chars = _virtualCharService.TryConvertToVirtualChars(token);
```

Comment:
`_virtualCharService` is only used here. I think it could be passed as parameter instead.
One of the two callers is an unused method. The other caller can easily pass that service through.

Reply:
Same reason as above.  i would prefer the only things passed in to be the information that varies.  The data tthat is fixed (i.e. the semantic model, services, etc.) i would like to all be kept together and only be provided once.